core/src/traits/Kokkos_OccupancyControlTrait.hpp

```diff
@@ -0,0 +1,209 @@
+// <editor-fold desc="Occupancy control user interface"> {{{1
+
+struct MaximizeOccupancy;
+
+struct DesiredOccupancy {
+  int m_occ = 100;
+  explicit constexpr DesiredOccupancy(int occ) : m_occ(occ) {
+    KOKKOS_EXPECTS(0 <= occ && occ <= 100);
+  }
+  explicit constexpr operator int() const { return m_occ; }
+  constexpr int value() const { return m_occ; }
+  DesiredOccupancy() = default;
+  explicit DesiredOccupancy(MaximizeOccupancy const&) : DesiredOccupancy() {}
+};
+
+struct MaximizeOccupancy {
+  explicit MaximizeOccupancy() = default;
+};
+
+// </editor-fold> end Occupancy control user interface }}}1
+//==============================================================================
+
+}  // end namespace Experimental
+
+namespace Impl {
+
+//==============================================================================
+// <editor-fold desc="Occupancy control trait specification"> {{{1
+
+struct OccupancyControlTrait : TraitSpecificationBase<OccupancyControlTrait> {
+  // MSVC workaround for linearizing base classes (see Impl::linearize_bases)
+  template <template <class> class GetBase, class... OtherTraits>
+  struct base_traits : linearize_bases<GetBase, OtherTraits...> {
+    using occupancy_control = Kokkos::Experimental::MaximizeOccupancy;
+    static constexpr bool experimental_contains_desired_occupancy = false;
+    // Default access occupancy_control, for when it is the (stateless) default
+    static constexpr occupancy_control impl_get_occupancy_control() {
+      return occupancy_control{};
+    }
+  };
+  template <class T>
+  using trait_matches_specification = std::integral_constant<
+      bool,
+      std::is_same<T, Kokkos::Experimental::DesiredOccupancy>::value ||
+          std::is_same<T, Kokkos::Experimental::MaximizeOccupancy>::value>;
+};
+
+// </editor-fold> end Occupancy control trait specification }}}1
+//==============================================================================
+
+//==============================================================================
+// <editor-fold desc="AnalyzeExecPolicy specializations"> {{{1
+
+// The DesiredOccupancy case has runtime storage, so we need to handle copies
+// and assignments
+template <class... Traits>
+struct AnalyzeExecPolicy<void, Kokkos::Experimental::DesiredOccupancy,
+                         Traits...> : AnalyzeExecPolicy<void, Traits...> {
+ public:
+  using base_t            = AnalyzeExecPolicy<void, Traits...>;
+  using occupancy_control = Kokkos::Experimental::DesiredOccupancy;
+  static constexpr bool experimental_contains_desired_occupancy = true;
+
+  template <class OccControl>
+  using with_occupancy_control = AnalyzeExecPolicy<void, OccControl, Traits...>;
+
+  // Treat this as private, but make it public so that MSVC will still treat
+  // this as a standard layout class and make it the right size: storage for a
+  // stateful desired occupancy
+  //   private:
+  occupancy_control m_desired_occupancy;
```

Comment:
maybe
```suggestion
  // Treat this as private, but make it public so that MSVC will still treat
  // this as a standard layout class and make it the right size: storage for a
  // stateful desired occupancy
  occupancy_control m_private_desired_occupancy;
```
then?

Reply:
Let's make this change in #3832 so that the rebase isn't as messy. I'm not sure this still needs to be the case in the new version for MSVC, so we should double check that.